File: 150-evaluate-reverse-polish-notation/150-evaluate-reverse-polish-notation.py
# Approach: Evaluate with Stack
# Time Complexity : O(n)
# Space Complexity : O(n)

class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        
        stack = [];
        
        for token in tokens:
            if token not in "+-*/":
                stack.append(int(token))
                continue
                
            num2 = stack.pop()
            num1 = stack.pop()
            
            res = 0
            if token == "+":
                res = num1 + num2
            elif token == "-":
                res = num1 - num2
            elif token == "*":
                res = num1 * num2
            else:
                res = int(num1 / num2)         
            stack.append(res)
        
        return stack.pop()
            
        

File: 150-evaluate-reverse-polish-notation/150-evaluate-reverse-polish-notation.js
/**
 * @param {string[]} tokens
 * @return {number}
 */

// convert a string containing a number into a number use +num
var evalRPN = function(tokens) {
    
    const OPERATORS = ["+", "-", "/", "*"];
    const stack = [];

    for (const token of tokens) {
        if (OPERATORS.includes(token)) {
            const number2 = stack.pop();
            const number1 = stack.pop();
            stack.push(operation(number1, number2, token));
        } else {
            stack.push(Number(token));
        }
    }

    return stack.pop();
};

var operation = function(num1, num2, token) {
    if (token === "+") {
        return num1 + num2; 
    } 
    if (token === "-") {
        return num1 - num2;
    }
    if (token === "*") {
        return num1 * num2;
    }
    if (token === "/") {
        return Math.trunc(num1 / num2);
    } 
}